spark/src/main/scala/org/apache/spark/sql/delta/commands/merge/ClassicMergeExecutor.scala

```diff
@@ -199,6 +210,43 @@ trait ClassicMergeExecutor extends MergeOutputGeneration {
     (touchedAddFiles, dedupe)
   }
 
+  /**
+   * Helper function that produces an expression by combining a sequence of clauses with OR.
+   * Requires the sequence to be non-empty.
+   */
+  protected def combineClausesWithOr(clauses: Seq[DeltaMergeIntoClause]): Expression = {
+    require(clauses.nonEmpty)
+    clauses
+      .map(_.condition.getOrElse(Literal.TrueLiteral))
+      .reduce((a, b) => Or(a, b))
+  }
+
+  protected def generateFilterForModifiedAndNewRows(
```

Comment:
Instead of having a flag `includeNotMatchedFilter`, you could split this method into `filterForModifiedRows` and `filterForInsertedRows` and use `Or(filterForModifiedRows, filterForInsertedRows)` when needing both